jupyterhub/tests/selenium/test_browser.py

```diff
@@ -854,3 +857,148 @@ async def test_user_logout(app, browser, url, user):
+        )
+    # verify that user can see the service name and oauth URL
+    text_permission = browser.find_element(
+        By.XPATH, './/h1[text()="Authorize access"]//following::p'
+    ).text
+    assert f"JupyterHub service {service.name}", (
+        f"oauth URL: {expected_redirect_url}" in text_permission
+    )
+    # permissions check
+    oauth_form = browser.find_element(By.TAG_NAME, "form")
+    scopes_elements = oauth_form.find_elements(
+        By.XPATH, '//input[@type="hidden" and @name="scopes"]'
+    )
+    scope_list_oauth_page = []
+    for scopes_element in scopes_elements:
+        # checking that scopes are invisible on the page
+        assert not scopes_element.is_displayed()
+        scope_value = scopes_element.get_attribute("value")
+        scope_list_oauth_page.append(scope_value)
+
+    # checking that all scopes granded to user are presented in POST form (scope_list)
+    assert all(x in scope_list_oauth_page for x in user_scopes)
+    assert f"access:services!service={service.name}" in scope_list_oauth_page
+
+    check_boxes = oauth_form.find_elements(
+        By.XPATH, '//input[@type="checkbox" and @name="raw-scopes"]'
+    )
+    for check_box in check_boxes:
+        # checking that user cannot uncheck the checkbox
+        assert not check_box.is_enabled()
+        assert check_box.get_attribute("disabled")
+        assert check_box.get_attribute("title") == "This authorization is required"
+
+    # checking that appropriete descriptions are displayed depending of scopes
+    descriptions = oauth_form.find_elements(By.TAG_NAME, 'span')
+    desc_list_UI = [
+        description.get_attribute("innerHTML")
+        .strip()
+        .replace('  ', '')
+        .replace('\n', '')
+        for description in descriptions
+    ]
+    # getting descriptions from scopes.py to compare them with descriptions on UI
+    scope_def = scopes.describe_raw_scopes(user_scopes, user.name)
+    list_of_desc_for_comparing = []
+    for scope_def in scope_def:
+        if "description" in scope_def:
+            text_desc_scope = scope_def.get("description")
+            text_filter = scope_def.get("filter")
+            if text_filter == "":
+                list_of_desc_for_comparing.append(text_desc_scope)
+            elif text_filter != "":
+                list_of_desc_for_comparing.append(
+                    text_desc_scope + f"Applies to {text_filter}."
+                )
+    assert sorted(desc_list_UI) == sorted(list_of_desc_for_comparing)
+
+    # click on the Authorize button
+    await click(browser, (By.XPATH, '//input[@type="submit"]'))
+    # check that user returned to service page
+    assert browser.current_url == service_url
+
+    # check the scope on the service page
+    # getting the scopes from the service page
+    text = browser.find_element(By.TAG_NAME, "body").get_attribute("innerHTML")
+    obj = json.loads(text)
+    list_of_scopes_service_page = obj["scopes"]
```

Comment:
simplifying getting the user model from the service API:

```suggestion
    # check the granted permissions by
    # getting the scopes from the service page,
    # which contains the JupyterHub user model
    text = browser.find_element(By.TAG_NAME, "body").text
    user_model = json.loads(text)
    authorized_scopes = user_model["scopes"]
```